docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,546 @@
+| ``+w:123``             | fixed-sized list [123 items]                      |            |
++------------------------+---------------------------------------------------+------------+
+| ``+s``                 | struct                                            |            |
++------------------------+---------------------------------------------------+------------+
+| ``+m``                 | map                                               | \(2)       |
++------------------------+---------------------------------------------------+------------+
+| ``+u:I,J,...``         | union with type ids I,J...                        |            |
++------------------------+---------------------------------------------------+------------+
+
+Notes:
+
+(1)
+   The timezone string is appended as-is after the colon character ``:``, without
+   any quotes.
+
+(2)
+   As specified in the Arrow columnar format, the map type has a single child type,
+   itself a 2-child struct type of ``(key, value)``.
+
+Examples
+--------
+
+* A dictionary-encoded ``decimal128`` array with ``int16`` indices has format
+  string ``s``, and its dependent dictionary array has format string ``d``.
+* A ``list<uint64>`` array has format string ``+l``, and its single child
+  has format string ``L``.
+* A ``struct<ints: int32, floats: float32>`` has format string ``+s``; its two
+  children have names ``ints`` and ``floats``, and format strings ``i`` and
+  ``f`` respectively.
+* A ``map<string, float64>`` array has format string ``+m``; its single child
+  has name ``entry`` and format string ``+s``; its two grandchildren have names
+  ``key`` and ``value``, and format strings ``u`` and ``g`` respectively.
+* A ``union<ints: int32, floats: float32>`` with type ids ``4, 5`` has format
+  string ``+u:4,5``; its two children have names ``ints`` and ``floats``,
+  and format strings ``i`` and ``f`` respectively.
+
+
+Structure definitions
+=====================
+
+The following free-standing definitions are enough to support the Arrow
+C data protocol in your project.  Like the rest of the Arrow project, they
+are available under the Apache License.
+
+.. code-block:: c
+
+   #define ARROW_FLAG_ORDERED 1
+   #define ARROW_FLAG_NULLABLE 2
+
+   struct ArrowArray {
+     // Type description
+     const char* format;
+     const char* name;
+     const char* metadata;
+     int32_t flags;
+
+     // Data description
+     int64_t length;
+     int64_t null_count;
+     int64_t offset;
+     int64_t n_buffers;
+     int64_t n_children;
+     const void** buffers;
+     struct ArrowArray** children;
+     struct ArrowArray* dictionary;
+
+     // Release callback
+     void (*release)(struct ArrowArray*);
+     // Opaque producer-specific data
+     void* private_data;
+   };
+
+The ArrowArray structure
+------------------------
+
+.. c:member:: const char* ArrowArray.format
+
+   Mandatory.  A null-terminated, UTF8-encoded string describing
```

Comment:
isn't null termination incompatible with UTF-8?

Reply:
No. It just forbids using the unicode character 0 :-)

Reply:
Just reread the spec you are right as always :)